ANMN/burst_averaged_product/burstproduct/burst_average_product.py

```diff
@@ -0,0 +1,401 @@
+
+def get_var_val_var_qc_exclusion(netcdf_file_path, varname, qc_flag):
+    """
+    for a qc flag values [0:9], returns the var_values array, and var_qc_flag_exclusion
+    which is a boolean array returning True for all qc >= qc_flag
+
+    Also exlude all TIME values outside of deployment range as defined in gatts
+    """
+    netcdf_file_obj       = Dataset(netcdf_file_path, 'r')
+    shape_var_array = netcdf_file_obj.variables[varname].shape
+
+    if len(shape_var_array) == 3:
+        # WQM files
+        var_values            = netcdf_file_obj.variables[varname][:, 0, 0]
+        var_qc_flag_exclusion = netcdf_file_obj.variables['%s_quality_control' % varname][:, 0, 0] >= qc_flag
+    elif len(shape_var_array) == 1:
+        # CTD files
+        var_values            = netcdf_file_obj.variables[varname][:]
+        var_qc_flag_exclusion = netcdf_file_obj.variables['%s_quality_control' % varname][:] >= qc_flag
+
+    # we exclude as well ALL TIMES before date_deployment_start, and after
+    # date_deployment_end
+    time                  = netcdf_file_obj.variables['TIME']
+    date_deployment_start = date2num(datetime.strptime(netcdf_file_obj.time_deployment_start, '%Y-%m-%dT%H:%M:%SZ'), time.units, time.calendar)
+    date_deployment_end   = date2num(datetime.strptime(netcdf_file_obj.time_deployment_end, '%Y-%m-%dT%H:%M:%SZ'), time.units, time.calendar)
+
+    time_before = date_deployment_start >= time[:]
+    time_after  = date_deployment_end <= time[:]
+
+    var_qc_flag_exclusion = [a or b for a, b in zip(time_before, var_qc_flag_exclusion)]
+    var_qc_flag_exclusion = [a or b for a, b in zip(time_after, var_qc_flag_exclusion)]
+
+    netcdf_file_obj.close()
+    return var_values, var_qc_flag_exclusion
+
+def list_var_not_to_average(netcdf_file_path):
+    """
+    return a list of variable to create a burst average for
+    """
+    netcdf_file_obj = Dataset(netcdf_file_path, 'r')
+    var_list = netcdf_file_obj.variables.keys()
+    var_list = [x for x in var_list if not x.endswith('_quality_control')]
+
+    for var_to_remove in ['TIME', 'LATITUDE', 'LONGITUDE', 'NOMINAL_DEPTH', 'VOLT', 'TIMESERIES', 'SSPD', 'CNDC']:
+        if var_to_remove in var_list:
+            var_list.remove(var_to_remove)
+
+    netcdf_file_obj.close()
+    return var_list
+
+def burst_average_data(time_values, var_values, burst_duration, var_qc_exclusion):
+    """
+    create burst average data for var_values. This is the core of the script
+    """
+    n_seconds_day = 24 * 3600
+    difft         = np.round(np.diff(time_values) * n_seconds_day) # this gives us a spike for each new burst
+    allowed_gap   = burst_duration
+    fill_value    = float('nan')
+
+    # look for start index of each burst, ie the index of each new spike of
+    # difft variable
+    idx_spike = [0] # initialise first index
+    for idx, val_boolean in enumerate(difft < allowed_gap):
+        if not val_boolean:
+            idx_spike.append(idx + 1)
```

Comment:
This is also something you don't need to repeat for each variable, as it only depends on the timestamps.

Also, I find this a bit more straightforward:
```python
spikes = np.where(difft >= allowed_gap)[0] + 1
idx_spike.extend(spikes)
```